docs/question/server.md

```diff
@@ -1,17 +1,104 @@
+
+<!-- prettier-ignore -->
+```html title="question.html"
+<pl-question-panel>
+  If $x = {{params.x}}$, what is $y$ if $y$ is double $x$?
+</pl-question-panel>
+<pl-number-input answers-name="y" label="$y =$"></pl-number-input>
+<pl-submission-panel>
+  {{feedback.y}}
+</pl-submission-panel>
+```
+
+More details about the `{{params.x}}` and `{{feedback.y}}` Mustache syntax can be found in the [`question.html` documentation](./html.md).
+
+### Step 1: `generate`
+
+First, the `generate` function is called to generate random parameters for the variant, and the correct answers. It should set `data["params"]` with the parameters for the question, and `data["correct_answers"]` with the correct answers. The parameters can then be used in the `question.html` file by using `{{params.NAME}}`.
+
+```python title="server.py"
+import random
+
+def generate(data):
+    # Generate random parameters for the question and store them in the data["params"] dict:
+    data["params"]["x"] = random.randint(5, 10)
+
+    # Also compute the correct answer (if there is one) and store in the data["correct_answers"] dict:
+    data["correct_answers"]["y"] = 2 * data["params"]["x"]
+```
+
+!!! info
+
+    In general, each function of the question generation process runs *after* all elements in the question. For example, the `generate()` function in `server.py` runs after all elements have run their `generate()` functions. This is important to remember when using the `data` dictionary, as it will contain the results *after* the elements have finished processing.
+
+### Step 2: `prepare`
+
+Next, the `prepare` function is called after all elements (e.g. `<pl-number-input>`) have run `generate()`. This is typically done to do any sort of final post-processing, but is not commonly used.
+
+### Step 3: `render`
+
+Next, the `render(data, html)` function is called to render the question. You can use this function to override how the question is rendered. The render function expects two parameters, `data` and `html`, and should return a string of HTML. The HTML after the mustache template has been rendered is available through the `html` parameter. This is typically only used for more advanced questions, and we won't need it for this example.
+
+### Step 4: `parse`
+
+When a student submits their answer, the `parse` function is called to parse the submitted answers after the individual elements have parsed them. This function can be used to display more-specific format errors than the input elements or to parse the input differently. When our parse function runs, `<pl-number-input>` will have already parsed the submitted value as an integer, and display an error to the student if it was invalid. For this question, we will allow the student to only submit positive integers, so we will check for that and set a format error with `data["format_errors"]` if it is negative.
```

Comment:
I don't know if it may be a good idea to also expand the explanation (even if not applicable to the example) that parse can also be used parse the input in a question-specific way. In other words, parse is not just to check for errors, it can also be used to interpret the input.

Reply:
I expanded the example in 189fcab7d634a129d31bfb5c5fd1de4e2ee3cbd9 showing how you could modify the submitted answer in `parse`.